boards/adafruit-macropad/src/macros.rs

```diff
@@ -0,0 +1,209 @@
+        disp
+    }};
+}
+
+#[macro_export]
+macro_rules! macropad_pins {
+    ($pac:expr, $sio:expr) => {
+        $crate::Pins::new(
+            $pac.IO_BANK0,
+            $pac.PADS_BANK0,
+            $sio.gpio_bank0,
+            &mut $pac.RESETS,
+        )
+    };
+}
+
+#[macro_export]
+macro_rules! macropad_clocks {
+    ($pac:expr, $watchdog:expr) => {
+        $crate::init_clocks_and_plls(
+            $crate::XOSC_CRYSTAL_FREQ,
+            $pac.XOSC,
+            $pac.CLOCKS,
+            $pac.PLL_SYS,
+            $pac.PLL_USB,
+            &mut $pac.RESETS,
+            &mut $watchdog,
+        )
+        .ok()
+        .unwrap()
+    };
+}
+
+#[macro_export]
+macro_rules! macropad_rotary_encoder {
+    ($pins:expr) => {
+        $crate::Rotary::new(
+            $pins.encoder_rota.into_pull_up_input(),
+            $pins.encoder_rotb.into_pull_up_input(),
+        )
+    };
+}
+
+#[macro_export]
+macro_rules! macropad_keypad {
+    ($pins:expr) => {
+        $crate::KeysTwelve {
+            key1: $pins.key1.into_pull_up_input(),
+            key2: $pins.key2.into_pull_up_input(),
+            key3: $pins.key3.into_pull_up_input(),
+            key4: $pins.key4.into_pull_up_input(),
+            key5: $pins.key5.into_pull_up_input(),
+            key6: $pins.key6.into_pull_up_input(),
+            key7: $pins.key7.into_pull_up_input(),
+            key8: $pins.key8.into_pull_up_input(),
+            key9: $pins.key9.into_pull_up_input(),
+            key10: $pins.key10.into_pull_up_input(),
+            key11: $pins.key11.into_pull_up_input(),
+            key12: $pins.key12.into_pull_up_input(),
+        }
+    };
+}
+
+pub struct KeysTwelve {
+    pub key1: Pin<Gpio1, PullUpInput>,
+    pub key2: Pin<Gpio2, PullUpInput>,
+    pub key3: Pin<Gpio3, PullUpInput>,
+    pub key4: Pin<Gpio4, PullUpInput>,
+    pub key5: Pin<Gpio5, PullUpInput>,
+    pub key6: Pin<Gpio6, PullUpInput>,
+    pub key7: Pin<Gpio7, PullUpInput>,
+    pub key8: Pin<Gpio8, PullUpInput>,
+    pub key9: Pin<Gpio9, PullUpInput>,
+    pub key10: Pin<Gpio10, PullUpInput>,
+    pub key11: Pin<Gpio11, PullUpInput>,
+    pub key12: Pin<Gpio12, PullUpInput>,
+}
+
+impl KeysTwelve {
+    pub fn get_0based(&self, idx: i8) -> Option<&dyn InputPin<Error = Infallible>> {
+        self.get_1based(1 + idx)
+    }
+
+    pub fn get_1based(&self, idx: i8) -> Option<&dyn InputPin<Error = Infallible>> {
+        match idx {
+            1 => Some(&self.key1),
+            2 => Some(&self.key2),
+            3 => Some(&self.key3),
+            4 => Some(&self.key4),
+            5 => Some(&self.key5),
+            6 => Some(&self.key6),
+            7 => Some(&self.key7),
+            8 => Some(&self.key8),
+            9 => Some(&self.key9),
+            10 => Some(&self.key10),
+            11 => Some(&self.key11),
+            12 => Some(&self.key12),
+            _ => None,
+        }
+    }
+
+    pub fn array_0based(&self) -> [&dyn InputPin<Error = Infallible>; 12] {
```

Comment:
Doesn't really need the 0based suffix, there's no other base option and it's pretty obvious it's going to consume all the pins. Maybe name it `to_array` to match standard Rust naming conventions?

Also, since this moves the pins permanently out of `KeysTwelve`, do we want a way to convert from `[&dyn InputPin<Error = Infallible>; 12]` back to `KeysTwelve`?